src/traversals/iterators.jl

```diff
@@ -0,0 +1,130 @@
+"""
+    abstract type IteratorAlgorithm
+
+`IteratorAlgorithm` is an abstract type which specifies a particular algorithm to use when iterating through a graph.
+"""
+abstract type IteratorAlgorithm end
+
+
+"""
+    struct DFSIterator <: IteratorAlgorithm
+
+`DFSIterator` is a struct which specifies using depth-first traversal to iterate through a graph. A source node must be supplied to construct this iterator as `DFSIterator(g::AbstractGraph, source::Int)`.
+    
+# Examples
+```julia-repl
+julia> g = smallgraph(:house)
+{5, 6} undirected simple Int64 graph
+
+julia> for node in DFSIterator(g, 1)
+           display(node)
+       end
+1
+2
+4
+3
+5
+```
+"""
+struct DFSIterator <: IteratorAlgorithm
+    graph::AbstractGraph
+    source::Int
+end
+
+
+"""
+    struct BFSIterator <: IteratorAlgorithm
+
+`BFSIterator` is a struct which specifies using breadth-first traversal to iterate through a graph. A source node must be supplied to construct this iterator as `BFSIterator(g::AbstractGraph, source::Int)`.
+        
+# Examples
+```julia-repl
+julia> g = smallgraph(:house)
+{5, 6} undirected simple Int64 graph
+
+julia> for node in BFSIterator(g, 1)
+           display(node)
+       end
+1
+2
+3
+4
+5
+```
+"""
+struct BFSIterator <: IteratorAlgorithm
+    graph::AbstractGraph
+    source::Int
+end
+
+
+"""
+    abstract type AbstractGraphIteratorState
+
+`BasicGraphIteratorState` is an abstract type to hold the current state of iteration which is need for Julia's Base.iterate() function.
+"""
+abstract type AbstractGraphIteratorState end
+
+
+"""
+    mutable struct BasicGraphIteratorState
+
+`BasicGraphIteratorState` is a struct to hold the current state of iteration which is need for Julia's Base.iterate() function. It is a basic implementation used for depth-first or breadth-first iterators.
+"""
+mutable struct BasicGraphIteratorState <: AbstractGraphIteratorState
+    visited::BitArray
+    queue::Vector{Int}
+end
+
+
+"""
+    Base.iterate(t::Union{BFSIterator, DFSIterator})
+
+First iteration to visit each node for depth-first or breadth-first type iterators.
+"""
+function Base.iterate(t::Union{BFSIterator, DFSIterator})
+    visited = falses(nv(t.graph))
+    visited[t.source] = true
+    state = BasicGraphIteratorState(visited, [t.source])
+    return (t.source, state)
+end
+
+
+"""
+    Base.iterate(t::DFSIterator, state::BasicGraphIteratorState)
+
+Iterator to visit each node in a depth-first manner.
+"""
+function Base.iterate(t::DFSIterator, state::BasicGraphIteratorState)
+    while !isempty(state.queue)
```

Comment:
If I understand this correctly, if the graph is is not connected (or only weakly connected in the directed case), we might not visit all vertices? Is that on purpose?

Reply:
It's a good catch but I think it make sense to only visit vertices reachable from the source, and we could have a separate Iterator (with no source) that would call a DFS Iterator on unexplored vertices until all vertices are explored.

Reply:
The assumption when doing DFS or BFS is you only explore connected vertices, at least that's my understanding (and personal use case)